sphinxcontrib/redirects/__init__.py

```diff
@@ -27,24 +28,47 @@ def generate_redirects(app):
         app.info("Could not find redirects file at '%s'" % path)
         return
 
-    in_suffix = app.config.source_suffix[0]
+    in_suffix = app.config.source_suffix
+    if isinstance(in_suffix, list):
+        in_suffix = in_suffix[0]
+    if isinstance(in_suffix, dict):
+        app.info("app.config.source_suffix is a dictionary type. "
+                 "Defaulting source_suffix to '.rst'")
+        in_suffix = ".rst"
 
-    # TODO(stephenfin): Add support for DirectoryHTMLBuilder
-    if not type(app.builder) == builders.StandaloneHTMLBuilder:
-        app.warn("The 'sphinxcontib-redirects' plugin is only supported "
-                 "by the 'html' builder. Skipping...")
+    if type(app.builder) == linkcheckbuilders.CheckExternalLinksBuilder:
+        app.info("Detected 'linkcheck' builder in use so skipping generating redirects")
         return
 
+    if not (type(app.builder) == builders.StandaloneHTMLBuilder or type(app.builder) == builders.DirectoryHTMLBuilder):
+        app.warn("The 'sphinxcontib-redirects' plugin is only supported "
+                 "by the 'html' and 'dirhtml' builder, but you are using '%s'. Skipping..." % type(app.builder))
+
+    dirhtml = False
+    if type(app.builder) == builders.DirectoryHTMLBuilder:
+        dirhtml = True
+
     with open(path) as redirects:
         for line in redirects.readlines():
             from_path, to_path = line.rstrip().split(' ')
 
             app.debug("Redirecting '%s' to '%s'" % (from_path, to_path))
 
-            from_path = from_path.replace(in_suffix, '.html')
+            if dirhtml:
+                from_path = from_path.replace(in_suffix, '/index.html')
+            else:
+                from_path = from_path.replace(in_suffix, '.html')
```

Comment:
It's hardcode, because we can change file extensions by [file_suffix](https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-html_file_suffix)

Actually I think that we can use `app.builder.get_target_uri` to get path